WIP Add testing DB2 to GitHub Actions

bahusoid opened the pull request:
Too many tests error..

hazzik committed:
Register truncate function for DB2

bahusoid committed:
Add testing DB2 to GitHub Actions

bahusoid committed:
DB2 dialect: Support str function

bahusoid committed:
Add DB2TestDialect

hazzik commented:
DB2 requires similar to Firebird workaround for parameters. Parameters that are used in the SELECT statements should be wrapped into `cast`.


```
2021-12-31T14:59:15.4008216Z 14) Error : NHibernate.Test.Criteria.ProjectionsTestAsync.UsingConditionalsAsync
2021-12-31T14:59:15.4010114Z NHibernate.Exceptions.GenericADOException : could not execute query
2021-12-31T14:59:15.4011188Z [ SELECT (case when this_.studentId = @p0 then @p1 else @p2 end) as y0_ FROM Student this_ ]
2021-12-31T14:59:15.4012866Z   Name:cp0 - Value:27 [Type: NHibernate.Type.Int64Type (SqlType: Int64)]  Name:cp1 - Value:yes [Type: NHibernate.Type.StringType (SqlType: String)]  Name:cp2 - Value:no [Type: NHibernate.Type.StringType (SqlType: String)]
2021-12-31T14:59:15.4014355Z [SQL: SELECT (case when this_.studentId = @p0 then @p1 else @p2 end) as y0_ FROM Student this_]
2021-12-31T14:59:15.4016006Z   ----> IBM.Data.DB2.Core.DB2Exception : ERROR [42610] [IBM][DB2/LINUXX8664] SQL0418N  The statement was not processed because the statement contains an invalid use of one of the following: an untyped parameter marker, the DEFAULT keyword, or a null value.
2021-12-31T14:59:15.4017111Z Data:
2021-12-31T14:59:15.4017905Z   actual-sql-query: SELECT (case when this_.studentId = @p0 then @p1 else @p2 end) as y0_ FROM Student this_
```

See this: https://www.ibm.com/docs/es/db2/11.1?topic=design-parameters-markers#d97366e93

hazzik commented:
I think we need to have a systematic approach rather than trying to fix this problem for different DBs. (I think Hana has the same problem)

bahusoid commented:
> DB2 requires similar to Firebird

I don't really like how it's done in Firebird. We need to think about more elegant solution rather than "brute force"  parameters wrapping in cast. The request to disable this functionality proves it can cause more harm than good

hazzik commented:
Yes, agree.

hazzik committed:
SupportsNonDataBoundCondition false

bahusoid commented:
```C#
171) Error : NHibernate.Test.TypesTest.StringTypeWithLengthFixture.HqlEqualityParameterCanExceedColumnSize
NHibernate.Exceptions.GenericADOException : could not execute query
[ select stringclas0_.Id as id1_0_, stringclas0_.StringValue as stringvalue2_0_, stringclas0_.LongStringValue as longstringvalue3_0_ from StringClass stringclas0_ where stringclas0_.StringValue=@p0 ]
  Name:likeValue - Value:AAAAAAAAABx [Type: NHibernate.Type.StringType (SqlType: String(Length=10))]
  ----> IBM.Data.DB2.Core.DB2Exception : ERROR [22001] [IBM] CLI0109E  String data right truncation. SQLSTATE=22001
```
This test is skipped if `SupportsNonDataBoundCondition` is false. But I see no non data bound conditions. It seems we need to adjust parameter sizes similar how it's done for MSSQL in `SqlClientDriver.AdjustParameterForValue`

hazzik commented:
> This test is skipped if SupportsNonDataBoundCondition is false.

Skip condition is incorrect then.

hazzik committed:
SupportsSubSelectsInOrderBy false

hazzik commented on src/NHibernate.Test/TestDialect.cs:
```diff
@@ -120,10 +120,15 @@ public bool SupportsSqlType(SqlType sqlType)
+		/// <summary>
+		/// Supports sub-selects in order by clause
+		/// </summary>
+		public virtual bool SupportsSubSelectsInOrderBy => _dialect.SupportsScalarSubSelects;
```
DB2 supports scalar sub selects, but it does not support them in order-by clause.

bahusoid replied:
And in SELECT statement seems too

hazzik replied:
@bahusoid this test works just fine:

https://github.com/nhibernate/nhibernate-core/blob/6576a9675cbfc48b560c58f85794013b4512131d/src/NHibernate.Test/Linq/SelectionTests.cs#L408-L464

hazzik committed:
Revert MySql condition

hazzik committed:
HasBrokenTypeInferenceOnSelectedParameters  true

hazzik committed:
SupportsHavingOnGroupedByComputation false

ERROR [42803] [IBM][DB2/LINUXX8664] SQL0119N  An expression starting with "BODYWEIGHT" specified in a SELECT clause, HAVING clause, or ORDER BY clause is not specified in the GROUP BY clause or it is in a SELECT clause, HAVING clause, or ORDER BY clause with a column function and no GROUP BY clause is specified.

hazzik committed:
SupportsSelectForUpdateWithPaging false

hazzik committed:
Simplifying exports for DB2

hazzik committed:
Add packages for all supported platforms.

hazzik committed:
Use 3.1.0.500 version of provider

hazzik committed:
DB2 precision for decimal seems to be 0(?)

hazzik committed:
DB2 reports int parameters with precision

hazzik committed:
Fix casing for column name

hazzik committed:
Skip native guid tests for DB2

hazzik commented:
>  Use NET5.IBM.Data.DB2 lib instead of IBM.Data.DB2.Core 

I was thinking about this. But why?

hazzik commented:
Also... The assembly name and namespace has reverted back to .NET version. Need to change driver.

bahusoid commented:
> I was thinking about this. But why?

Just to check how it works... It's the latest version. So why not?

bahusoid commented:
> Just to check how it works.

And it doesn't work... So let's just drop it for now..

hazzik commented:
Yeah, it is `IBM.Data.Db2.DB2Connection` (in assembly) vs `IBM.Data.DB2.DB2Connection` (expected)

bahusoid commented:
Changing casing in namespace. Smart move...

hazzik committed:
Fix tests wit multi-table delete

hazzik committed:
Add QuerySequencesString

hazzik committed:
Register replace function

hazzik committed:
Register reserved keywords

bahusoid commented:
Interesting async regen failure.
<details> 
  <summary>Logs </summary>
crit: AsyncGenerator.CommandLine[0]
      One or more errors occurred while opening the project:
      Msbuild failed when processing the file '/home/runner/work/nhibernate-core/nhibernate-core/src/NHibernate.Test/NHibernate.Test.csproj' with message: /usr/share/dotnet/sdk/6.0.405/Microsoft.Common.CurrentVersion.targets: (2302, 5): There was a mismatch between the processor architecture of the project being built "MSIL" and the processor architecture of the reference "/home/runner/.nuget/packages/ibm.data.db2.core-lnx/3.1.0.500/lib/netstandard2.1/IBM.Data.DB2.Core.dll", "AMD64". This mismatch may cause runtime failures. Please consider changing the targeted processor architecture of your project through the Configuration Manager so as to align the processor architectures between your project and references, or take a dependency on references with a processor architecture that matches the targeted processor architecture of your project.
      Hint: For suppressing irrelevant errors use SuppressDiagnosticFailures option.
      System.InvalidOperationException: One or more errors occurred while opening the project:
      Msbuild failed when processing the file '/home/runner/work/nhibernate-core/nhibernate-core/src/NHibernate.Test/NHibernate.Test.csproj' with message: /usr/share/dotnet/sdk/6.0.405/Microsoft.Common.CurrentVersion.targets: (2302, 5): There was a mismatch between the processor architecture of the project being built "MSIL" and the processor architecture of the reference "/home/runner/.nuget/packages/ibm.data.db2.core-lnx/3.1.0.500/lib/netstandard2.1/IBM.Data.DB2.Core.dll", "AMD64". This mismatch may cause runtime failures. Please consider changing the targeted processor architecture of your project through the Configuration Manager so as to align the processor architectures between your project and references, or take a dependency on references with a processor architecture that matches the targeted processor architecture of your project.
      Hint: For suppressing irrelevant errors use SuppressDiagnosticFailures option.
         at AsyncGenerator.AsyncCodeGenerator.CheckForErrors(MSBuildWorkspace workspace, String itemType, ImmutableArray`1 supressFailuresPredicates, ILogger logger) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\AsyncCodeGenerator.cs:line 437
         at AsyncGenerator.AsyncCodeGenerator.OpenProject(MSBuildWorkspace workspace, String filePath, ImmutableArray`1 supressFailuresPredicates, ILogger logger, CancellationToken cancellationToken) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\AsyncCodeGenerator.cs:line 375
         at AsyncGenerator.AsyncCodeGenerator.GenerateAsync(AsyncCodeConfiguration configuration, CancellationToken cancellationToken) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator\AsyncCodeGenerator.cs:line 76
         at AsyncGenerator.CommandLine.Program.Main(String[] args) in C:\Workspace\Git\AsyncGenerator\Source\AsyncGenerator.CommandLine\Program.cs:line 51
</details>

robsosno commented:
> Changing casing in namespace. Smart move...

I've just moved my app from .NET 3.1 to .NET 6.0. Next step it would be to have new Db2 driver supporting .NET 6.0 (I expect better performance). However I was surprised with driver name change - [db2-net-packages](https://community.ibm.com/community/user/datamanagement/blogs/vishwa-hs1/2020/07/12/db2-net-packages-download-and-configure) (breaking changes paragraph).
I see some logic in the naming: IBM.Data.DB2 is dedicated to classic .NET framework, IBM.Data.DB2.Core to .NET Core and IBM.Data.Db2 to .NET 5+ . They removed Core similarly as Framework did and to avoid possibility of incorrect linking 3.1 with 5+ they've changed DB2 to Db2. So it seems that all future versions will be in accordance with this new schema.
I saw some changes on [db2Net5](https://github.com/bahusoid/nhibernate-core/tree/db2Net5/src/NHibernate/Driver) branch. When do you plan to merge it to the core?
I would vote to implement driver differently - as an additional class, let it be named Db2Driver or Db2Driver5. I trust more to manual configuration than smart switches.

bahusoid committed:
Move exports to DB_INIT

bahusoid committed:
Use latest docker db2:11.5.8.0

hazzik committed:
Use .NET6/7 version of provider

hazzik committed:
Ignore MSB3270 for AsyncGenerator

hazzik committed:
Add new driver for Net5.IBM.Data.Db2/Net.IBM.Data.Db2 DataProvider

hazzik committed:
Update NHibernate.Test.csproj

hazzik committed:
use 11.5.9 & net8

hazzik committed:
Set SupportsCancelQuery to false

hazzik committed:
DB2NetDriver should be same as DB2CoreDriver

hazzik committed:
DB2 has "id" as a reserved keyword :-/

hazzik committed:
Fix ignoring test

hazzik commented:
Note to myself:

On Windows need to set `DB2LANG=en_US` environment variable.

github-actions[bot] committed:
Generate async files

hazzik committed:
Add new DB2NetDriver and obsolete DB2CoreDriver

hazzik committed:
Merge branch 'master' of https://github.com/nhibernate/nhibernate-core into db2

hazzik committed:
Merge branch 'db2netdriver' into db2

# Conflicts:
#	src/NHibernate/Driver/DB2NetDriver.cs

hazzik committed:
Merge branch 'db2' of https://github.com/bahusoid/nhibernate-core into db2

hazzik committed:
Fix package name

hazzik committed:
Register "strguid" function

hazzik committed:
Register some math functions

hazzik committed:
Register some string functions

hazzik committed:
Register nvl & iif functions

hazzik committed:
fixup! Register some string functions

hazzik committed:
DB2 does not support sequence identity

github-actions[bot] committed:
Generate async files

hazzik committed:
Register atan2 function

hazzik committed:
Add support for temporary tables

github-actions[bot] committed:
Generate async files

hazzik committed:
DB2 supports union all

hazzik committed:
Update reserved words list based on documentation

hazzik commented:
Current state: 91 tests are failing.

hazzik committed:
Revert "DB2 has "id" as a reserved keyword :-/"

This reverts commit 578bf11b1aa29accab18fe03000272abb9dc6d2c.
